rcl_yaml_param_parser/test/test_yaml_variant.cpp

```diff
@@ -0,0 +1,140 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+
+#include <yaml.h>
+
+#include <string>
+#include <vector>
+
+#include "osrf_testing_tools_cpp/scope_exit.hpp"
+#include "rcl_yaml_param_parser/parser.h"
+#include "rcl_yaml_param_parser/impl/yaml_variant.h"
+#include "rcutils/allocator.h"
+#include "rcutils/strdup.h"
+#include "rcutils/testing/fault_injection.h"
+
+#define TEST_VARIANT_COPY(dest_variant, src_variant, field, tmp_var, allocator) \
+  do { \
+    SCOPED_TRACE("TEST_VARIANT_COPY " #field); \
+    src_variant.field = &tmp_var; \
+    EXPECT_TRUE(rcl_yaml_variant_copy(&dest_variant, &src_variant, allocator)); \
+    ASSERT_NE(nullptr, dest_variant.field); \
+    EXPECT_EQ(*src_variant.field, *dest_variant.field); \
+    rcl_yaml_variant_fini(&dest_variant, allocator); \
+    variant.field = nullptr; \
+  } while (0)
+
+#define TEST_VARIANT_ARRAY_COPY( \
+    dest_variant, src_variant, field, array_type, value_type, tmp_array, array_size, allocator) \
+  do { \
+    SCOPED_TRACE("TEST_VARIANT_ARRAY_COPY " #field); \
+    src_variant.field = \
+      static_cast<array_type *>(allocator.allocate(sizeof(array_type), allocator.state)); \
+    ASSERT_NE(nullptr, src_variant.field); \
+    src_variant.field->values = \
+      static_cast<value_type *>( \
+      allocator.zero_allocate(array_size, sizeof(value_type), allocator.state)); \
+    ASSERT_NE(nullptr, src_variant.field->values); \
+    src_variant.field->size = array_size; \
+    for (size_t i = 0; i < array_size; ++i) { \
+      variant.field->values[i] = tmp_array[i]; \
+    } \
+    EXPECT_TRUE(rcl_yaml_variant_copy(&dest_variant, &src_variant, allocator)); \
+    ASSERT_NE(nullptr, dest_variant.field); \
+    ASSERT_NE(nullptr, dest_variant.field->values); \
+    for (size_t i = 0; i < array_size; ++i) { \
+      SCOPED_TRACE(i); \
```

Comment:
What does this macro do? I didn't know it. I suppose it's part of the logging of the test, so this won't block the PR.

Reply:
Yah, it's a helpful gtest macro for loops, where it's hard to tell during which iteration a check might have failed. By using `i`, I can print out the loop count that a check fails. Combined with the SCOPED_TRACE in line 43 I can also print out the field type. For example, if I add an `EXPECT_EQ(1,0)` here the failure message looks like:

```
/home/brawner/workspace/ros2_fi/src/ros2/rcl/rcl_yaml_param_parser/test/test_yaml_variant.cpp:100: Failure
6: Expected equality of these values:
6:   1
6:   0
6: Google Test trace:
6: /home/brawner/workspace/ros2_fi/src/ros2/rcl/rcl_yaml_param_parser/test/test_yaml_variant.cpp:100: 0
6: /home/brawner/workspace/ros2_fi/src/ros2/rcl/rcl_yaml_param_parser/test/test_yaml_variant.cpp:100: TEST_VARIANT_ARRAY_COPY bool_array_value
6: /home/brawner/workspace/ros2_fi/src/ros2/rcl/rcl_yaml_param_parser/test/test_yaml_variant.cpp:100: Failure
....
6: Google Test trace:
6: /home/brawner/workspace/ros2_fi/src/ros2/rcl/rcl_yaml_param_parser/test/test_yaml_variant.cpp:100: 1
6: /home/brawner/workspace/ros2_fi/src/ros2/rcl/rcl_yaml_param_parser/test/test_yaml_variant.cpp:100: TEST_VARIANT_ARRAY_COPY bool_array_value
...
6: /home/brawner/workspace/ros2_fi/src/ros2/rcl/rcl_yaml_param_parser/test/test_yaml_variant.cpp:104: Failure
...
6: Google Test trace:
6: /home/brawner/workspace/ros2_fi/src/ros2/rcl/rcl_yaml_param_parser/test/test_yaml_variant.cpp:104: 0
6: /home/brawner/workspace/ros2_fi/src/ros2/rcl/rcl_yaml_param_parser/test/test_yaml_variant.cpp:104: TEST_VARIANT_ARRAY_COPY integer_array_value
6: /home/brawner/workspace/ros2_fi/src/ros2/rcl/rcl_yaml_param_parser/test/test_yaml_variant.cpp:104: Failure
...
```

Reply:
Thanks for the explanation!